app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt

```diff
@@ -203,6 +205,63 @@ constructor(
     }
   }
 
+  private fun observeEvents() {
+    viewModelScope.launch {
+      eventRepository.observeEvents().collect { resource ->
+        when (resource) {
+          is Resource.Success -> {
+            _uiState.update { currentState ->
+              currentState.copy(
+                  eventList =
+                      EventList(
+                          resource.data, resource.data, _uiState.value.eventList.selectedEvent))
+            }
+            filterEvents()
+          }
+          is Resource.Failure ->
+              Log.d("EventsOverviewViewModel", "Failed to fetch events: ${resource.throwable}")
+        }
+      }
+    }
+  }
+
+  private fun observeUpcomingEvents() {
+    viewModelScope.launch {
+      userRepository.getCurrentUserId().let { userIdResource ->
+        when (userIdResource) {
+          is Resource.Success -> {
+            val uid = userIdResource.data
+            eventRepository.observeUpcomingEvents(uid).collect { resource ->
+              when (resource) {
+                is Resource.Success -> {
+                  _uiState.update { currentState ->
+                    currentState.copy(
+                        upcomingEventList =
+                            EventList(
+                                allEvents = resource.data,
+                                filteredEvents = resource.data,
+                                selectedEvent = currentState.upcomingEventList.selectedEvent))
+                  }
+                  filterEvents()
+                }
+                is Resource.Failure -> {
+                  Log.d(
+                      "EventsOverviewViewModel",
+                      "Failed to fetch upcoming events: ${resource.throwable.message}")
+                }
+              }
+            }
+          }
+          is Resource.Failure -> {
+            Log.d(
+                "EventsOverviewViewModel",
+                "Error fetching user ID: ${userIdResource.throwable.message}")
+          }
+        }
+      }
+    }
+  }
```

Comment:
There is *a lot* of indentation in this code, up to **12** which probably sets a new record! ^^ I'd suggest reducing the indentation using modularization. It will be way easier to read and understand the flow of your code. Here is an example:

```suggestion
private fun observeUpcomingEvents() {
    viewModelScope.launch {
        val userIdResource = userRepository.getCurrentUserId()
        handleUserIdResource(userIdResource)
    }
}

private suspend fun handleUserIdResource(userIdResource: Resource<String>) {
    when (userIdResource) {
        is Resource.Success -> {
            val uid = userIdResource.data
            val eventsResource = eventRepository.observeUpcomingEvents(uid).collect()
            handleEventsResource(eventsResource)
        }
        is Resource.Failure -> {
            Log.d(
                "EventsOverviewViewModel",
                "Error fetching user ID: ${userIdResource.throwable.message}"
            )
        }
    }
}

private suspend fun handleEventsResource(eventsResource: Resource<List<Event>>) {
    when (eventsResource) {
        is Resource.Success -> {
            updateUiState(eventsResource.data)
            filterEvents()
        }
        is Resource.Failure -> {
            Log.d(
                "EventsOverviewViewModel",
                "Failed to fetch upcoming events: ${eventsResource.throwable.message}"
            )
        }
    }
}

private fun updateUiState(events: List<Event>) {
    _uiState.update { currentState ->
        currentState.copy(
            upcomingEventList = EventList(
                allEvents = events,
                filteredEvents = events,
                selectedEvent = currentState.upcomingEventList.selectedEvent
            )
        )
    }
}
```

Which version would you say is easier to read/follow ?

Reply:
Your suggestion is a lot easier to understand. Thank you!